### PR TITLE
KEYCLOAK-5657 Support for transient NameIDPolicy and AllowCreate in S…

### PR DESCRIPTION
**Pull request by cgeorgilakis**
Related PR to keycloak-documentation :  https://github.com/keycloak/keycloak-documentation/pull/1117

See my comment in jira issue.

### CODE REVIEW

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java [hunk 1]
```diff
@@ -130,6 +130,9 @@ public Response performLogin(AuthenticationRequest request) {
                 requestedAuthnContext.addAuthnContextDeclRef(authnContextDeclRef);
 
             String loginHint = getConfig().isLoginHint() ? request.getAuthenticationSession().getClientNote(OIDCLoginProtocol.LOGIN_HINT_PARAM) : null;
+            Boolean allowCreate = null;
+            if ( getConfig().getConfig().get("allowCreate")== null ||  getConfig().isAllowCreate())
+                allowCreate =  Boolean.TRUE ;
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:135]:
```suggestion
                allowCreate = Boolean.TRUE ;
```

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java [hunk 2]
```diff
@@ -130,6 +130,9 @@ public Response performLogin(AuthenticationRequest request) {
                 requestedAuthnContext.addAuthnContextDeclRef(authnContextDeclRef);
 
             String loginHint = getConfig().isLoginHint() ? request.getAuthenticationSession().getClientNote(OIDCLoginProtocol.LOGIN_HINT_PARAM) : null;
+            Boolean allowCreate = null;
+            if ( getConfig().getConfig().get("allowCreate")== null ||  getConfig().isAllowCreate())
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:134]: Prefer constants to literals:
```suggestion
            if (getConfig().getConfig().get(ALLOW_CREATE) == null || getConfig().isAllowCreate())
```

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderConfig.java [hunk 3]
```diff
@@ -334,12 +336,24 @@ public boolean isSignSpMetadata() {
...
+    }
 
     @Override
     public void validate(RealmModel realm) {
         SslRequired sslRequired = realm.getSslRequired();
 
         checkUrl(sslRequired, getSingleLogoutServiceUrl(), SINGLE_LOGOUT_SERVICE_URL);
         checkUrl(sslRequired, getSingleSignOnServiceUrl(), SINGLE_SIGN_ON_SERVICE_URL);
+        //transient name id format is not accepted together with principaltype SubjectnameId
+        if ( JBossSAMLURIConstants.NAMEID_FORMAT_TRANSIENT.get().equals(getConfig().get(NAME_ID_POLICY_FORMAT)) && SamlPrincipalType.SUBJECT.name().equals(getConfig().get(PRINCIPAL_TYPE)))
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderConfig.java:355]:
```suggestion
        if (JBossSAMLURIConstants.NAMEID_FORMAT_TRANSIENT.get().equals(getNameIDPolicyFormat()) && SamlPrincipalType.SUBJECT == getPrincipalType()) 
```

File: themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js [hunk 4]
```diff
@@ -935,17 +935,17 @@ module.controller('RealmIdentityProviderCtrl', function($scope, $filter, $upload
...
     $scope.initSamlProvider = function() {
         $scope.nameIdFormats = [
-            /*
-            {
-                format: "urn:oasis:names:tc:SAML:2.0:nameid-format:transient",
-                name: "Transient"
-            },
-            */
+         
+
```
Comment (hmlnarik) [inline themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js:939]: Remove the empty lines.
Also at the end of this `initSamlProvider` function, set the default value for `allowCreate` flag, something like:
```javascript
        $scope.identityProvider.config.allowCreate = 'true';
```
This should be `true` for the time being to stay backwards-compatible. Changing it to false may be done in the future.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlIdPInitiatedSsoTest.java [hunk 5]
```diff
@@ -396,6 +401,109 @@ public void testProviderIdpInitiatedLoginWithPrincipalAttribute() throws Excepti
...
+        assertEquals(1, userList.size());
+        String id = userList.get(0).getId();
+        FederatedIdentityRepresentation fed = users.get(id).getFederatedIdentity().get(0);
+        assertThat(fed.getUserId(), is(PROVIDER_REALM_USER_NAME));
+        assertThat(fed.getUserName(), is(PROVIDER_REALM_USER_NAME));
+        
+        //check that no user with sent subject-id was sent
+        userList= users.search("subjectId1");
+        assertTrue(userList.isEmpty());
+        userList= users.search("subjectId2");
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlIdPInitiatedSsoTest.java:504]:
```suggestion
        userList = users.search("subjectId1");
        assertTrue(userList.isEmpty());
        userList = users.search("subjectId2");
```

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Thank you for the PR!
Please see comments inline. After fixing few minor issues, this PR can go in.

File: themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js (hunk 4 above)
Comment (cgeorgilakis) [reply themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js:939]: I set default value to false in order to be compatible with default SAML value (Section 3.4.1.1 in https://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf). However, I will change it as you desired.

**Comment by cgeorgilakis** [comment]
I have done these changes.

File: themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js (hunk 4 above)
Comment (hmlnarik) [reply themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js]: You are indeed right about the specification. The reason for defaulting to `true` is to stay conformant to the current behaviour. In the future, the default might be changed, and that would impact also the default that is set here:
https://github.com/keycloak/keycloak/blob/809ab5bf002f0f34781e27be7cbfb29bcef793bd/services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java#L134-L135

File: themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js [hunk 6]
```diff
@@ -1039,6 +1037,7 @@ module.controller('RealmIdentityProviderCtrl', function($scope, $filter, $upload
         $scope.identityProvider.firstBrokerLoginFlowAlias = 'first broker login';
         $scope.identityProvider.config.useJwksUrl = 'true';
         $scope.identityProvider.config.syncMode = 'IMPORT';
+        $scope.identityProvider.config.allowCreate = 'true';
```
Comment (hmlnarik) [inline themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js:1040]: This should be _at the end of `initSamlProvider` function_, i.e. line 1013.

File: themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js (hunk 4 above)
Comment (hmlnarik) [reply themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js]: Unresolving this conversation as this comment needs to be addressed at the proper place

**Review by hmlnarik — APPROVED** [review]
Thank you for the contribution!

File: themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js (hunk 6 above)
Comment (cgeorgilakis) [reply themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js]: I have one question based on this change. 
I put default value only for create. 
When IdP was updated, we do not want to change allowCreate value. However, when upgrading from older version allowCreate  will not be in databse. So, when admin user try to update SAML Identity Provider value will be set to false. Is this a problem?
We do not have problem, when user is trying to login ( default to true).

**Pull request merged**